Mathlib/Analysis/Meromorphic/Divisor/Basic.lean

```diff
@@ -0,0 +1,420 @@
+Authors: Stefan Kebekus
+-/
+import Mathlib.Analysis.Normed.Field.Basic
+import Mathlib.Analysis.SpecialFunctions.Log.Basic
+
+/-!
+# Divisors on subsets of normed fields
+
+This file defines divisors, a standard book-keeping tool in complex analysis
+used to keep track of pole/vanishing orders of meromorphic objects, typically
+functions or differential forms. It provides supporting API and establishes
+divisors as an instance of a lattice ordered commutative group.
+
+Throughout the present file, `𝕜` denotes a nontrivially normed field, and `U` a
+subset of `𝕜`.
+
+## TODOs
+
+- Constructions: The divisor of a meromorphic function, behavior under product
+  of meromorphic functions, behavior under addition, behavior under restriction
+- Construction: The divisor of a rational polynomial
+-/
+
+open Filter Metric Real Set Topology
+
+variable {𝕜 : Type*} [NontriviallyNormedField 𝕜] {U : Set 𝕜}
+
+/-!
+## Definition, coercion to functions and basic extensionality lemmas
+
+A divisor on `U` is a function `𝕜 → ℤ` whose support is discrete within `U` and
+entirely contained within `U`.  The theorem
+`supportDiscreteWithin_iff_locallyFiniteWithin` shows that this is equivalent to
+the textbook definition, which requires the support of `f` to be locally finite
+within `U`.
+-/
+
+/-- A divisor on `U` is a triple specified below. -/
+structure DivisorOn (U : Set 𝕜) where
+  /-- A function `𝕜 → ℤ` -/
+  toFun : 𝕜 → ℤ
+  /-- A proof that the support of `toFun` is contained in `U` -/
+  supportWithinDomain' : toFun.support ⊆ U
+  /-- A proof the the support is discrete within `U` -/
+  supportDiscreteWithinDomain' : toFun =ᶠ[codiscreteWithin U] 0
+
+/-- A divisor is a divisor on `⊤ : Set 𝕜`. -/
+def Divisor (𝕜 : Type*) [NontriviallyNormedField 𝕜] := DivisorOn (⊤ : Set 𝕜)
+
+/-- The condition `supportDiscreteWithinU` in a divisor is equivalent to saying
+that the support is locally finite near every point of `U`. -/
+theorem supportDiscreteWithin_iff_locallyFiniteWithin {f : 𝕜 → ℤ} (h : f.support ⊆ U) :
+    f =ᶠ[codiscreteWithin U] 0 ↔ ∀ z ∈ U, ∃ t ∈ 𝓝 z, Set.Finite (t ∩ f.support) := by
+  have : f.support = (U \ {x | f x = (0 : 𝕜 → ℤ) x}) := by
+    ext x
+    simp only [Function.mem_support, ne_eq, Pi.zero_apply, mem_diff, mem_setOf_eq, iff_and_self]
+    exact (h ·)
+  rw [EventuallyEq, Filter.Eventually, codiscreteWithin_iff_locallyFiniteComplementWithin, this]
+
+namespace DivisorOn
+
+/-- Divisors are `FunLike`: the coercion from divisors to functions is injective. -/
+instance : FunLike (DivisorOn U) 𝕜 ℤ where
+  coe D := D.toFun
+  coe_injective' := fun ⟨_, _, _⟩ ⟨_, _, _⟩ ↦ by simp
+
+/-- This allows writing `D.support` instead of `Function.support D` -/
+abbrev support (D : DivisorOn U)  := Function.support D
+
+lemma supportWithinDomain (D : DivisorOn U) : D.support ⊆ U := D.supportWithinDomain'
+
+lemma supportDiscreteWithinDomain (D : DivisorOn U) : D =ᶠ[codiscreteWithin U] 0 :=
+  D.supportDiscreteWithinDomain'
```

Comment:
Here we state the fields of the structure again, with the original names, but the statement now uses the coercion, instead of `toFun`.
So, when using these lemmas, `toFun` doesn't suddenly appear in the goal state.